portfolio/src/main/java/com/google/sps/servlets/TranslateServlet.java

```diff
@@ -0,0 +1,81 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.cloud.translate.Translate;
+import com.google.cloud.translate.TranslateOptions;
+import com.google.cloud.translate.Translation;
+import com.google.gson.Gson;
+import com.google.sps.data.TranslationData;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.HashMap;
+import java.util.Locale;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that translates text for languages in a majority of countries around the world. */
+@WebServlet("/translate")
+public class TranslateServlet extends HttpServlet {
+  private String text; // Text to be translated to each language
+  private HashMap<String, String> translationCache; // Stores translation for each language
```

Comment:
The cache is for the languages because many countries share languages (i.e. English/Spanish/Chinese) and the translate API takes a little bit of time to translate each time so I thought I could reduce the amount of time the overall translation takes by saving the results for each language and checking if the text has already been translated to that language before passing it to the API as I go through the languages of each country.

Reply:
So my question is why aren't the keys for the cache both the language and the text.

Right now the key is only the language, so how do you know the value is for the right text?

Reply:
This servlet is called once for each time the user puts text into the website. It translates the same word/phase into multiple languages around the world.

Reply:
Hmm, maybe I'm missing something here then. 

So it seems like the cache is just being used just for one translate call (and not for any calls later on as they can be for a different text). Then doesn't it make more sense to either make the cache not a class variable, and use it just for this specific call, or to have the cache take in two keys (language, text).

On a side note, is it not possible to batch the translation calls instead of going language by language?

Reply:
You're right, and I made the cache not be a class variable anymore to reflect this. I am planning to look into batch translation later today.